File: backend/internal/events/bus.go
package events

import "sync"

type AppEvent struct {
	ID             int64  `json:"id,omitempty"`
	Type           string `json:"type"`
	IssueID        int64  `json:"issueId,omitempty"`
	ProjectID      string `json:"projectId,omitempty"`
	ThreadID       string `json:"threadId,omitempty"`
	TurnID         string `json:"turnId,omitempty"`
	Method         string `json:"method,omitempty"`
	AgentEventType string `json:"agent_event_type,omitempty"`
	Provider       string `json:"provider,omitempty"`
	RawMethod      string `json:"raw_method,omitempty"`
	RawPayload     string `json:"raw_payload,omitempty"`
	Command        string `json:"command,omitempty"`
	Path           string `json:"path,omitempty"`
	Status         string `json:"status,omitempty"`
	Text           string `json:"text,omitempty"`
	Error          string `json:"error,omitempty"`
	Payload        string `json:"payload,omitempty"`
	CreatedAt      string `json:"created_at,omitempty"`
}

type Bus struct {
	mu   sync.Mutex
	next int
	subs map[int]chan AppEvent
}

func NewBus() *Bus {
	return &Bus{subs: map[int]chan AppEvent{}}
}

func (b *Bus) Subscribe() (<-chan AppEvent, func()) {
	b.mu.Lock()
	defer b.mu.Unlock()
	id := b.next
	b.next++
	ch := make(chan AppEvent, 64)
	b.subs[id] = ch
	return ch, func() { b.unsubscribe(id) }
}

func (b *Bus) Publish(event AppEvent) {
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, ch := range b.subs {
		select {
		case ch <- event:
		default:
		}
	}
}

func (b *Bus) unsubscribe(id int) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if ch, ok := b.subs[id]; ok {
		delete(b.subs, id)
		close(ch)
	}
}

File: backend-ts/src/events/bus.ts
const BUS_BUFFER_SIZE = 64;

export type AppEvent = {
  id?: number;
  type: string;
  issueId?: number;
  conversationId?: string;
  projectId?: string;
  threadId?: string;
  turnId?: string;
  method?: string;
  agent_event_type?: string;
  provider?: string;
  raw_method?: string;
  raw_payload?: string;
  command?: string;
  path?: string;
  status?: string;
  text?: string;
  error?: string;
  payload?: string;
  created_at?: string;
};

type Subscriber = {
  events: AppEvent[];
  notify?: () => void;
};

type EventHandler = (event: AppEvent) => void;

export class EventBus {
  #nextID = 0;
  #handlers = new Set<EventHandler>();
  #subscribers = new Map<number, Subscriber>();

  publish(event: AppEvent): void {
    for (const handler of this.#handlers) handler(event);
    for (const subscriber of this.#subscribers.values()) {
      subscriber.events.push(event);
      if (subscriber.events.length > BUS_BUFFER_SIZE) subscriber.events.shift();
      subscriber.notify?.();
      subscriber.notify = undefined;
    }
  }

  subscribe(): EventSubscription {
    const id = this.#nextID;
    this.#nextID += 1;
    this.#subscribers.set(id, { events: [] });
    return new EventSubscription(() => this.#subscribers.get(id), () => this.#subscribers.delete(id));
  }

  subscriberCount(): number {
    return this.#subscribers.size;
  }

  observe(handler: EventHandler): () => void {
    this.#handlers.add(handler);
    return () => {
      this.#handlers.delete(handler);
    };
  }
}

export class EventSubscription {
  #closed = false;

  constructor(
    private readonly getSubscriber: () => Subscriber | undefined,
    private readonly unsubscribe: () => boolean
  ) {}

  close(): void {
    if (this.#closed) return;
    const subscriber = this.getSubscriber();
    this.#closed = true;
    subscriber?.notify?.();
    if (subscriber) subscriber.notify = undefined;
    this.unsubscribe();
  }

  async next(): Promise<AppEvent | undefined> {
    const subscriber = this.getSubscriber();
    if (!subscriber || this.#closed) return undefined;
    const event = subscriber.events.shift();
    if (event) return event;
    return await new Promise((resolve) => {
      subscriber.notify = () => resolve(subscriber.events.shift());
    });
  }
}
